pkg/cloud/azure/azure_storage.go

```diff
@@ -125,13 +153,41 @@ func parseAzureURL(
 		return conf, errors.Errorf("azure uri missing %q parameter", AzureAccountNameParam)
 	}
 
-	hasKeyCreds := conf.AzureConfig.AccountKey != ""
-
-	hasRoleCreds := conf.AzureConfig.TenantID != "" && conf.AzureConfig.ClientID != "" && conf.AzureConfig.ClientSecret != ""
-	noRoleCreds := conf.AzureConfig.TenantID == "" && conf.AzureConfig.ClientID == "" && conf.AzureConfig.ClientSecret == ""
-
-	if hasRoleCreds == hasKeyCreds || hasRoleCreds == noRoleCreds {
-		return conf, errors.Errorf("azure uri requires exactly one authentication method: %q OR all three of %q, %q, and %q", AzureAccountKeyParam, AzureTenantIDParam, AzureClientIDParam, AzureClientSecretParam)
+	// Validate the authentication parameters are set correctly.
+	switch conf.AzureConfig.Auth {
+	case cloudpb.ExternalStorage_Azure_AUTH_LEGACY:
+		hasKeyCred := conf.AzureConfig.AccountKey != ""
+		hasARoleCred := conf.AzureConfig.TenantID != "" || conf.AzureConfig.ClientID != "" || conf.AzureConfig.ClientSecret != ""
+		if !hasKeyCred || hasARoleCred {
+			// If the URI params are misconfigured we can't be certain Legacy Auth
+			// was intended, so print a broader error message.
+			return conf, errors.Errorf("explicit azure uri requires exactly one authentication method: %q OR all three of %q, %q, and %q", AzureAccountKeyParam, AzureTenantIDParam, AzureClientIDParam, AzureClientSecretParam)
```

Comment:
nit: declare the error message above the switch stmt to avoid writing it out twice?